libraries/AP_HAL_PX4/CAN.cpp

```diff
@@ -0,0 +1,1035 @@
+/*
+ * Copyright (C) 2014 Pavel Kirienko <pavel.kirienko@gmail.com>
+ *
+ * With modifications for Ardupilot CAN driver
+ * Copyright (C) 2017 Eugene Shamaev
+ */
+
+//#define CONFIG_HAL_BOARD HAL_BOARD_PX4
+#include <AP_HAL/AP_HAL.h>
+#include <AP_HAL/system.h>
+
+#include <AP_BoardConfig/AP_BoardConfig.h>
+
+#include <cassert>
+#include <cstring>
+#include "CAN.h"
+
+#include <nuttx/irq.h>
+#include <unistd.h>
+#include <drivers/drv_hrt.h>
+
+#include <assert.h>
+
+#include "Scheduler.h"
+
+/*
+ * FOR INVESTIGATION:
+ * AP_HAL::micros64() was called for monotonic time counter
+ * pavel-kirienko: This will work as long as we don't need to synchronize the autopilot's own clock with an external
+ * time base, e.g. a GNSS time provided by an external GNSS receiver. Libuavcan's STM32 driver supports automatic time
+ * synchronization only if it has a dedicated hardware timer to work with.
```

Comment:
@OXINARF @EShamaev @tridge This will become highly relevant once we'll want to use the same synchronized time domain shared across all nodes on the bus. Zubax GNSS can serve as a time synchronization master, synchronizing the clocks of all nodes with the global UTC time base with microsecond resolution. The theory is explained here: http://uavcan.org/Specification/6._Application_level_functions/#time-synchronization; the algorithm is based on the concepts elaborated in this paper: “Implementing a Distributed High-Resolution Real-Time Clock using the CAN-Bus” (M. Gergeleit and H. Streich).